src/operations/drop.ts

```diff
@@ -24,10 +31,84 @@ export class DropCollectionOperation extends CommandOperation<boolean> {
     session: ClientSession | undefined,
     callback: Callback<boolean>
   ): void {
-    super.executeCommand(server, session, { drop: this.name }, (err, result) => {
-      if (err) return callback(err);
-      if (result.ok) return callback(undefined, true);
-      callback(undefined, false);
+    (async () => {
+      const db = this.db;
+      const options = this.options;
+      const name = this.name;
+
+      const encryptedFieldsMap = db.s.client.options.autoEncryption?.encryptedFieldsMap;
+      let encryptedFields: Document | undefined =
+        options.encryptedFields ?? encryptedFieldsMap?.[`${db.databaseName}.${name}`];
+
+      if (!encryptedFields && encryptedFieldsMap) {
+        // If the MongoClient was configued with an encryptedFieldsMap,
+        // and no encryptedFields config was available in it or explicitly
+        // passed as an argument, the spec tells us to look one up using
+        // listCollections().
+        const listCollectionsResult = await db
+          .listCollections({ name }, { nameOnly: false })
+          .toArray();
+        encryptedFields = listCollectionsResult?.[0]?.options?.encryptedFields;
+      }
+
+      let result;
+      let errorForMainOperation;
+      try {
+        result = await this.executeWithoutEncryptedFieldsCheck(server, session);
+      } catch (err) {
+        if (
+          !encryptedFields ||
+          (err instanceof MongoServerError && err.code !== MONGODB_ERROR_CODES.NamespaceNotFound)
+        ) {
+          throw err;
+        }
+        // Save a possible NamespaceNotFound error for later
+        // in the encryptedFields case, so that the auxilliary
+        // collections will still be dropped.
+        errorForMainOperation = err;
+      }
+
+      if (encryptedFields) {
+        const escCollection = encryptedFields.escCollection || `enxcol_.${name}.esc`;
+        const eccCollection = encryptedFields.eccCollection || `enxcol_.${name}.ecc`;
+        const ecocCollection = encryptedFields.ecocCollection || `enxcol_.${name}.ecoc`;
+
+        for (const collectionName of [escCollection, eccCollection, ecocCollection]) {
+          // Drop auxilliary collections, ignoring potential NamespaceNotFound errors.
+          const dropOp = new DropCollectionOperation(db, collectionName);
+          try {
+            await dropOp.executeWithoutEncryptedFieldsCheck(server, session);
+          } catch (err) {
```

Comment:
are we intentionally swallowing any errors that might not be instances of MongoServerError?

Reply:
ah yeah no, thanks for noticing – messed this up during code review in 75c02b70ee573232737dbc727c998c628ad1cc0a, is fixed now :slightly_smiling_face: